ctapipe/reco/shower_processor.py

```diff
@@ -0,0 +1,194 @@
+    Needs DL1_PARAMETERS as input.
+    Should be run after ImageProcessor to produce all DL1b information.
+
+    For the moment it only supports the reconstruction of the shower geometry
+    using ctapipe.reco.HillasReconstructor.
+
+    It is planned to support also energy reconstruction and particle type
+    classification which are now disabled by default.
+    """
+
+    reconstruct_energy = Bool(
+        default_value=False, help="Reconstruct the energy of the event."
+    ).tag(config=True)
+
+    classify = Bool(
+        default_value=False, help="Classify the particle type associated with the event."
+    ).tag(config=True)
+
+    def __init__(
+        self,
+        subarray: SubarrayDescription,
+        config=None,
+        parent=None,
+        **kwargs
+    ):
+        """
+        Parameters
+        ----------
+        subarray: SubarrayDescription
+            Description of the subarray. Provides information about the
+            camera which are useful in calibration. Also required for
+            configuring the TelescopeParameter traitlets.
+        config: traitlets.loader.Config
+            Configuration specified by config file or cmdline arguments.
+            Used to set traitlet values.
+            This is mutually exclusive with passing a ``parent``.
+        parent: ctapipe.core.Component or ctapipe.core.Tool
+            Parent of this component in the configuration hierarchy,
+            this is mutually exclusive with passing ``config``
+        """
+
+        super().__init__(config=config, parent=parent, **kwargs)
+        self.subarray = subarray
+        self.check_shower = ShowerQualityQuery(parent=self)
+        self.reconstructor = HillasReconstructor()
+
+    def _reconstruct_shower(
+        self,
+        event,
+        default=DEFAULT_SHOWER_PARAMETERS,
+    ) -> ReconstructedShowerContainer:
+        """Perform shower reconstruction.
+
+        Parameters
+        ----------
+        event : container
+            A `ctapipe` event container
+        default: container
+            The default 'ReconstructedShowerContainer' which is
+            filled with NaNs.
+        Returns
+        -------
+        ReconstructedShowerContainer:
+            direction in the sky with uncertainty,
+            core position on the ground with uncertainty,
+            h_max with uncertainty,
+            is_valid boolean for successfull reconstruction,
+            average intensity of the intensities used for reconstruction,
+            measure of algorithm success (if fit),
+            list of tel_ids used if stereo, or None if Mono
+        """
+
+        # Select only images which pass the shower quality criteria
+        hillas_dict = {
+            tel_id: dl1.parameters.hillas
+            for tel_id, dl1 in event.dl1.tel.items()
+            if all(self.check_shower(dl1.parameters))
+        }
+        self.log.debug(
+            "shower_criteria:\n %s", self.check_shower.to_table()
```

Comment:
Just because you changed how this is formatted, `to_table` is still a pretty expensive operation that is executed here everytime

Reply:
Indeed, but we are proposing here to not input this to the logger, but rather write it directly to output

Reply:
such an info is anyway too complex to make sense printing it to prompt

Reply:
if you really want to keep it for now, you can wrap it in:
```py3
if self.log.isEnabledFor(logging.DEBUG):
    self.log.debug(
            "shower_criteria:\n %s", self.check_shower.to_table()
```

That way it's not executed unless the log-level is DEBUG. Another option would be to add a __str__ method to QualityQuery that returns the table in string format, that way if you write:
```py3
self.log.debug("shower_criteria:\n %s", self.check_shower)
```
the table method it is only called if debug is on, and no if-statement is needed

Reply:
I like the second proposal

Reply:
done